### ISSUE TITLE
Future Plans & Direction

### ISSUE THREAD

**Issue by georgeaf99** [xuset/p2pweb#2]
Thread to be used for discussing future plans for the project, goals for the MVP, and a rough timeline.

**Comment by georgeaf99** [comment]
When we spoke the other day at Medici, we agreed that our goal was to create a distributed content creation platform where peers could submit "posts" which could be upvoted by their peers.

Breakdown:
- Create a broadcast system which allows peers to publish posts
- Allow peers to attach their identity to a post as an upvote
- Give peers the ability to verify the identities of other peers (each peer can only have one identity)

**Comment by xuset** [comment]
For the future plans and direction I see there being two main parts: a generic p2p lib using webrtc and post-based website that utilizes the p2p lib. The lib's goals should be to simplifying the storage, retrieval, and discovery of distributed and decentralized data. With the basic building blocks of storage, retrieval, and discovery, a p2p post-based website can be built on top of it. I believe there should be a clear separation of concerns in regards to the two facets; the lib should not concern itself with posts or upvotes, and instead only focus on arbitrary data.

I think these are the properties we should go for in the mvp:
 * Lib: webrtc for p2p communication
 * Lib: Be able to store, retrieve, and discover arbitrary and anonymously authored data.
 * Website: Ability to create and upvote posts.

In the longterm, having authors/identities for posts and comments will be nice. Fighting spam will be challenging, and may be curtailed be requiring a proof of work.

**Comment by georgeaf99** [comment]
I agree with everything you said. We will have to think about how to generalize the backend cleanly; perhaps we can use an existing distributed storage service and rebuild the peers to work using WebRTC.  Additionally, upvoting may have to be part of the lib if it requires a custom tailored algorithm.

**Comment by xuset** [comment]
What existing storage services could we use?

I agree that keeping upvoting out of the lib will be tricky but I'm hoping we can find a way to generalize it.

**Comment by georgeaf99** [comment]
If we are going to include a distributed data store as part of the lib, then we could use an existing lightweight distributed database and modify it to run over WebRTC.

**Comment by xuset** [comment]
That would be optimal, but I'm not sure we can find a db that fits our constraints. Every distributed database I know expects a very low churn rate for the nodes while our use case will have a very high churn rate since users are the nodes, and they can come and go as they please. If we run our own always-on nodes and don't treat the users as nodes in the system then we will have a low churn and can use an existing distributed db implementation, but I'm not a fan of that idea. If we can find a distributed db that works well under high churn we should definitely look into it. Building a distributed data store under high churn is a hard problem, but it's not unsolvable.

**Comment by georgeaf99** [comment]
**Summary of May 19th Discussion:**

Our product will be a "reddit clone" which allows users to create posts, discover posts, and upvote/comment on them. 

Each node will connect to a somewhere around 10-50 other nodes in the network and communicate with them over WebRTC. When a node wants to create a post it will submit it to its known peers and these nodes will replicate it to their known peers. Each node will keep a list of trackers with links/keys to specific posts in the system. When a node wants specific information for a given post it will look it up using the distributed hash table and then become one of the seeders for that post. When a node wants its peers to replicate its post or upvote it will send them a proof of work that is a combination of its public key and the post nonce. Each peer will require a proof of work proportional to the rate of posts coming from a specific node.
